### PR TITLE
Create Patient, inheriting from Person

### PR DESCRIPTION
**Pull request by iamjackslayer**
Patient model is now available. This class extends from Person. Currently it has preferredDoctor and appointment, I will add consultation and medicine soon (ideally tomorrow).

### CODE REVIEW

**Comment by iamjackslayer** [comment]
I will make those who urgently need to depend on patient (or rather Person for the time being) reviewers. @jjlee050 I assigned you to review to ensure smooth transition of the task (creating Patient).

**Comment by iamjackslayer** [comment]
I will make those who urgently need to depend on patient (or rather Person for the time being) reviewers. @jjlee050 I assigned you to review to ensure smooth transition of the task (creating Patient).

File: src/main/java/seedu/address/model/person/Person.java [hunk 1]
```diff
@@ -60,6 +82,41 @@ public Address getAddress() {
...
+    public void setAppointment(Appointment appointment) {
+        requireNonNull(appointment);
+
+        this.appointment = Optional.of(appointment);
+    }
+
+    /**
+     * Checks if the patient has preferred doctor.
+     */
+    public boolean hasPrefferedDoctor() {
```
Comment (gingivitiss) [inline src/main/java/seedu/address/model/person/Person.java]: there are some instances where preferred is spelt wrong.

File: src/main/java/seedu/address/model/person/Person.java [hunk 2]
```diff
@@ -23,6 +27,8 @@
     // Data fields
```
Comment (gingivitiss) [inline src/main/java/seedu/address/model/person/Person.java]: will you be adding consultation, medical and appointment history?

File: src/main/java/seedu/address/model/person/Person.java [hunk 3]
```diff
@@ -71,7 +128,9 @@ public boolean isSamePerson(Person otherPerson) {
 
         return otherPerson != null
                 && otherPerson.getName().equals(getName())
-                && (otherPerson.getPhone().equals(getPhone()) || otherPerson.getEmail().equals(getEmail()));
+                && (otherPerson.getPhone().equals(getPhone()) || otherPerson.getEmail().equals(getEmail()))
+                && otherPerson.getPrefferedDoctor().equals(getPrefferedDoctor())
```
Comment (jjlee050) [inline src/main/java/seedu/address/model/person/Person.java]: Spelling mistake on preferred doctor.

File: src/main/java/seedu/address/model/person/Person.java [hunk 4]
```diff
@@ -93,13 +152,15 @@ public boolean equals(Object other) {
...
+                && otherPerson.getTags().equals(getTags())
+                && otherPerson.getPrefferedDoctor().equals(getPrefferedDoctor())
+                && otherPerson.getAppointment().equals(getAppointment());
     }
 
     @Override
     public int hashCode() {
         // use this method for custom fields hashing instead of implementing your own
-        return Objects.hash(name, phone, email, address, tags);
+        return Objects.hash(name, phone, email, address, tags, getPrefferedDoctor(), getAppointment());
```
Comment (jjlee050) [inline src/main/java/seedu/address/model/person/Person.java]: Its quite weird here to have variable name here then suddenly got method name.

File: src/main/java/seedu/address/model/person/Person.java [hunk 5]
```diff
@@ -23,6 +27,8 @@
     // Data fields
     private final Address address;
     private final Set<Tag> tags = new HashSet<>();
+    private Optional<Doctor> preferredDoctor = Optional.empty();
```
Comment (jjlee050) [inline src/main/java/seedu/address/model/person/Person.java]: Not sure if its good to add here. Currently, my doctor extends Person (Person <---- Doctor). Does this mean that doctor have preferred doctor and appointment?

Comment (jjlee050) [reply src/main/java/seedu/address/model/person/Person.java]: and so is receptionist

File: src/main/java/seedu/address/model/person/Person.java [hunk 6]
```diff
@@ -52,6 +58,22 @@ public Address getAddress() {
         return address;
     }
 
+    /**
+     * Returns the patient's preferred doctor wrapped in {@link Optional}. The patient may not have one.
+     * @return an Optional {@link Doctor}.
+     */
+    public Optional<Doctor> getPrefferedDoctor() {
```
Comment (jjlee050) [inline src/main/java/seedu/address/model/person/Person.java]: Wrong spelling for preferred.

File: src/test/java/seedu/address/testutil/PersonBuilder.java [hunk 7]
```diff
@@ -44,6 +51,8 @@ public PersonBuilder(Person personToCopy) {
         email = personToCopy.getEmail();
         address = personToCopy.getAddress();
         tags = new HashSet<>(personToCopy.getTags());
+        preferredDoctor = personToCopy.getPrefferedDoctor();
```
Comment (jjlee050) [inline src/test/java/seedu/address/testutil/PersonBuilder.java]: Same as above

File: src/test/java/seedu/address/testutil/PersonBuilder.java [hunk 8]
```diff
@@ -86,8 +95,39 @@ public PersonBuilder withEmail(String email) {
...
     }
 
+    /**
+     * Contructs a Person based on the PersonBuilder.
+     * @return a person.
+     */
+    public Person build() {
+        Person person = new Person(name, phone, email, address, tags);
+
+        preferredDoctor.ifPresent(doctor -> {
```
Comment (jjlee050) [inline src/test/java/seedu/address/testutil/PersonBuilder.java]: @gingivitiss Is it better to put preferred doctor & appointment inside the new Person(..., here) or seperate?

File: src/main/java/seedu/address/model/person/Person.java [hunk 9]
```diff
@@ -23,6 +27,8 @@
     // Data fields
     private final Address address;
     private final Set<Tag> tags = new HashSet<>();
+    private Optional<Doctor> preferredDoctor = Optional.empty();
+    private Optional<Appointment> appointment = Optional.empty();
```
Comment (jjlee050) [inline src/main/java/seedu/address/model/person/Person.java]: I dont think you want to do initialisation here. Its better to do it inside constructor.

File: src/test/java/seedu/address/model/person/PersonTest.java [hunk 10]
```diff
@@ -56,6 +61,26 @@ public void isSamePerson() {
         // same name, same phone, same email, different attributes -> returns true
         editedAlice = new PersonBuilder(ALICE).withAddress(VALID_ADDRESS_BOB).withTags(VALID_TAG_HUSBAND).build();
         assertTrue(ALICE.isSamePerson(editedAlice));
+
```
Comment (jjlee050) [inline src/test/java/seedu/address/model/person/PersonTest.java]: What about pass test cases? Are u implementing TDD here?

File: src/main/java/seedu/address/model/person/Person.java (hunk 1 above)
Comment (iamjackslayer) [reply src/main/java/seedu/address/model/person/Person.java]: Oh I missed it, thanks!

File: src/main/java/seedu/address/model/person/Person.java (hunk 2 above)
Comment (iamjackslayer) [reply src/main/java/seedu/address/model/person/Person.java]: Yes, I will be adding soon. I will update you on this.

File: src/main/java/seedu/address/model/person/Person.java (hunk 3 above)
Comment (iamjackslayer) [reply src/main/java/seedu/address/model/person/Person.java]: noted with thanks.

File: src/main/java/seedu/address/model/person/Person.java (hunk 4 above)
Comment (iamjackslayer) [reply src/main/java/seedu/address/model/person/Person.java]: Noted. I will refactor it.

File: src/test/java/seedu/address/model/person/PersonTest.java (hunk 10 above)
Comment (iamjackslayer) [reply src/test/java/seedu/address/model/person/PersonTest.java]: Thanks for the heads up. I overlooked pass test cases. I think even with TDD I still need those pass cases at the same time (due to the implementation of Person#isSamePerson).

**Comment by iamjackslayer** [comment]
Need to merge or else we still don't have Patient

**Pull request merged**